remotes/docker/resolver.go

```diff
@@ -241,11 +275,18 @@ func (r *dockerResolver) Resolve(ctx context.Context, ref string) (string, ocisp
 			return "", ocispec.Descriptor{}, err
 		}
 
-		// turns out, we have a valid digest, make a url.
-		paths = append(paths, []string{"manifests", dgst.String()})
+		if immutableTag {
+			// Look up the blob by the immutable tag, not by the digest.
+			// The digest of the found blob is compared with the ref digest at the end of the function.
+			paths = append(paths, []string{"manifests", refTag})
+			caps |= HostCapabilityResolve
```

Comment:
> Enforcement of immutable tags on the client side would need to be done before by injecting the digest or after resolution by verifying the digest (same as signature verification).

The digest is verified in L423 after resolution:
``go
return "", ocispec.Descriptor{}, fmt.Errorf("immutable tag %q expects digest %q, got digest %q: %w", refTag, refDgst, dgst, ErrImmutableTagWasMutated)
```

Reply:
I see, it's just making sure they match. I still don't quite understand the feature then, since it is not necessary to complete the requested pull operation. I think it would be preferable to have single method for tag/digest verification if possible, but we should have a discussion about it.